File: P152996.py
def gcd(a, b):
    if b > a:
        a, b = b, a
    
    while b:
        a, b = b, a%b
    
    return a

def solution(weights):
    target = [2, 3, 4]
    answer = 0
    cand = dict()
    for weight in weights:
        cand[weight] = cand.get(weight, 0) + 1
    
    weights = []
    for key in cand:
        weights.append(key)
    size = len(weights)
    
    for x in range(size):
        for y in range(x, size):
            a, b = weights[x], weights[y]
            if a == b:
                answer += cand[a]*(cand[a]-1)//2
            else:
                for i in target:
                    for j in target:
                        if a*i == b*j:
                            answer += cand[a]*cand[b]
    return answer

File: P152996.java
class P152996 {
    public long solution(int[] weights) {
        long answer = 0;
        int[] cand = new int[1001];
        for (int weight : weights) cand[weight] += 1;
        int[] target = {2, 3, 4};
        
        for (int i=100; i<=1000; i++) {
            for (int j=i; j<=1000; j++) {
                if (i == j) answer += (long) cand[i]*(cand[i]-1)/2;
                else {
                    for (int a : target) {
                        for (int b : target) {
                            if (a*i == b*j) {
                                answer += (long) cand[i]*cand[j];   
                            }
                        }
                    }
                }
            }
        }
        return answer;
    }
}